### PR TITLE
fix: "Types in a binary operation should match, but found T and T"

### LINKED ISSUE #4635: Errors when comparing types
### Aim

I'm running into multiple issues related to type inference. This is one of them, but I'm not entirely sure how to title this report because I don't really know what might be the underlying issue. I produced a small exaple that triggers it instead.

Ultimately this errors out with `error: Types in a binary operation should match, but found T and T`, which just seems wrong.

### Expected Behavior

Forgive the boiler-plate, hopefully it won't be too bad. It's just some standard utils for dealing with Fields and arrays as is done in the kernel circuits.

```rust
trait FromField {
    fn from_field(field: Field) -> Self;
}

impl FromField for Field {
    fn from_field(value: Field) -> Self {
        value
    }
}

trait Deserialize<N> {
    fn deserialize(fields: [Field; N]) -> Self;
}

global AZTEC_ADDRESS_LENGTH = 1;

struct AztecAddress {
    inner : Field
}

impl FromField for AztecAddress {
    fn from_field(value: Field) -> Self {
        Self { inner: value }
    }
}

impl Deserialize<AZTEC_ADDRESS_LENGTH> for AztecAddress {
    fn deserialize(fields: [Field; AZTEC_ADDRESS_LENGTH]) -> Self {
        AztecAddress::from_field(fields[0])
    }
}

// Custom code

struct MyStruct<T> {
    a: T
}

impl<T> Deserialize<1> for MyStruct<T> {
    fn deserialize(fields: [Field; 1]) -> Self where T: FromField {
        Self{ a: FromField::from_field(fields[0]) }
    }
}

fn main() {
    let fields = [5; 1];
    let foo = MyStruct::deserialize(fields); // Note I don't specify T here (the type of `foo.a`)

    let bar = AztecAddress { inner: 5 };

    // Here `T` is apparently inferred to be `AztecAddress`, presumably because of the comparison.
    assert(foo.a == bar);
}
```

This errors out with:

```
error: Types in a binary operation should match, but found AztecAddress and AztecAddress
   ┌─ /home/nventuro/type_error/src/main.nr:49:12
   │
49 │     assert(foo.a == bar);
   │            ------------
```

which feels a bit non-sensical.

### Bug

Honestly I'm not sure if the issue is my code, the error message or the compiler, but something does seem off here.

### Installation Method

Compiled from source

### Nargo Version

nargo version = 0.25.0 noirc version = 0.25.0+7deceac68414a6c24f2c4f9ecdd36ad6ac9752e3 (git version hash: 7deceac68414a6c24f2c4f9ecdd36ad6ac9752e3, is dirty: false)

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves #4635 
Resolves https://github.com/noir-lang/noir/issues/4502

## Summary\*

This was more difficult to fix than it originally seemed. The main issue was between interactions with unbound type variables, type aliases, type rules for operators, and operator traits.

Removing the "infer unbound type variables to be numeric" rule on operators causes a lot of stdlib code to break where it'd be unreasonable to have type annotations. This caused unbound type variables to be bound to the first object type whose impl it was checked against when calling verify trait impl.

I eventually settled on just delaying the verify trait impl check for operators until the end of a function when more types are known.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by nventuro** [comment]
I just found a simpler scenario that also triggers this bug, in case you want to test against it or add it as a regression:

```rust
struct Wrapper {
    inner: Field,
}

trait FromField {
    fn from_field(value: Field) -> Self;
}

trait ToField {
    fn to_field(self) -> Field;
}

impl ToField for Wrapper {
    fn to_field(self) -> Field {
        self.inner
    }
}

impl FromField for Wrapper {
    fn from_field(value: Field) -> Wrapper {
        Wrapper { inner: value }
    }
}

#[test]
fn test() {
    let wapper = Wrapper { inner: 37 };
    assert_eq(FromField::from_field(wapper.to_field()), wapper);
}
```

This errors out with `Types in a binary operation should match, but found Wrapper and Wrapper`

**Comment by jfecher** [comment]
@nventuro I think the original example is sufficient. I'll mention though that the new error message with this PR for your snippet is:
```
error: No matching impl found for `Wrapper: Eq`
   ┌─ /.../short/src/main.nr:28:5
   │
28 │     assert_eq(FromField::from_field(wapper.to_field()), wapper);
   │     ----------------------------------------------------------- No impl for `Wrapper: Eq`
   │
```

**Comment by nventuro** [comment]
Ah yes, I didn't bother adding that impl for the minimum reproducible example :sweat_smile:

File: compiler/noirc_frontend/src/hir/type_check/expr.rs [hunk 1]
```diff
@@ -1079,36 +1081,42 @@ impl<'interner> TypeChecker<'interner> {
...
-        };
+        let use_impl = !lhs_type.is_numeric();
 
-        self.unify(lhs_type, &target, || TypeCheckError::TypeMismatchWithSource {
-            expected: lhs_type.clone(),
-            actual: rhs_type.clone(),
-            source: Source::Binary,
-            span,
-        });
+        // if the type variable is an integer or field we have to narrow it to only an integer
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/hir/type_check/expr.rs:1103]: this comment doesn't seem to account for the below condition `!op.kind.is_valid_for_field_type()`:

perhaps `if the type variable is an integer or field, but this binary operation only supports integers, we need to narrow it to only an integer`?

File: compiler/noirc_frontend/src/hir/type_check/expr.rs [hunk 2]
```diff
@@ -1079,36 +1081,42 @@ impl<'interner> TypeChecker<'interner> {
...
+        if !op.kind.is_valid_for_field_type() && lhs_type.is_numeric() {
+            // In addition to unifying both types, we also have to bind either
+            // the lhs or rhs to an integer type variable. This ensures if both lhs
+            // and rhs are type variables, that they will have the correct integer
+            // type variable kind instead of TypeVariableKind::Normal.
+            let target = Type::polymorphic_integer(self.interner);
+
+            self.unify(lhs_type, &target, || TypeCheckError::TypeMismatchWithSource {
+                expected: lhs_type.clone(),
+                actual: rhs_type.clone(),
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/hir/type_check/expr.rs:1113]: Should this be `actual: target`?

**Review by michaeljklein — CHANGES_REQUESTED** [review]

File: compiler/noirc_frontend/src/hir/type_check/expr.rs (hunk 2 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/hir/type_check/expr.rs]: I think the whole error message is wrong. I updated it to use the errors we already have for field comparisons, bitwise ops, and modulo.

**Pull request merged**